### PR TITLE
Modifies qMFKG.evaluate() to work with project, expand and cost aware utility

### LINKED ISSUE #587: [Bug] Issues with qKG.evaluate()
# 🐛 Bug

i) This line raises an `AttributeError` with models that do not have an `_input_batch_shape` attribute. (#578 as an example)
https://github.com/pytorch/botorch/blob/7e2a4040e60c6351ff127d6ff65506955c2beccb/botorch/optim/initializers.py#L328

ii) Since it is a subclass of `qKnowledgeGradient`, `qMultiFidelityKnowledgeGradient` has `evaluate` method implemented. However, the current implementation ignores the `project` operator in the inner optimization, and may produce buggy output. 

## Expected Behavior

i) The batch shape could be inferred from `model.train_inputs` batch shape. My observation is `_input_batch_shape = model.train_inputs[0].shape[:-2]`, but I am not sure if this is always true. If this is true, a simple try/except should do it.

ii) It should either raise a `NotImplementedError`, or be modified to accommodate the `project` operator.

### PR DESCRIPTION
**Pull request by saitcakmak**
## Motivation

Modifies `qMFKG.evaluate()` to work with `project`, `expand` and `cost_aware_utility`. Partially fixes #587. 

- Introduces a `ProjectedValueFunction` that wraps the `value_function` and applies the `project` operator on the `forward` call.
- Changes `evaluate()` signature to use `X` instead of `X_actual`. Current implementation raises an exception with the decorators when called with `evaluate(X_actual=...)`.

Note: The treatment of `cost_aware_utility` assumes that it is monotone non-decreasing in `deltas`. Otherwise, optimizing the inner problem and passing through `cost_aware_utility` may not produce the correct output.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added mock unit tests. Verified the expected behavior in additional offline tests.

### CODE REVIEW

File: botorch/acquisition/knowledge_gradient.py [hunk 1]
```diff
@@ -245,6 +253,10 @@ def evaluate(self, X_actual: Tensor, bounds: Tensor, **kwargs: Any) -> Tensor:
         if self.current_value is not None:
             values = values - self.current_value
 
+        if getattr(self, "cost_aware_utility", None) is not None:
```
Comment (Balandat) [inline botorch/acquisition/knowledge_gradient.py:256]:
```suggestion
        if hasattr(self, "cost_aware_utility"):
```

File: botorch/acquisition/knowledge_gradient.py [hunk 2]
```diff
@@ -428,16 +443,44 @@ def forward(self, X: Tensor) -> Tensor:
         return values.mean(dim=0)
 
 
+class ProjectedValueFunction(AcquisitionFunction):
+    r"""
+    Defines a wrapper around the value function that incorporates the project
+    operator.
```
Comment (Balandat) [inline botorch/acquisition/knowledge_gradient.py:449]:
```suggestion
    Defines a wrapper around  an `AcquisitionFunction` that incorporates the project
    operator. Typically used to handle value functions in look-ahead methods.
```

File: botorch/acquisition/knowledge_gradient.py [hunk 3]
```diff
@@ -428,16 +443,44 @@ def forward(self, X: Tensor) -> Tensor:
...
+    """
+
+    def __init__(
+        self,
+        base_value_function: AcquisitionFunction,
+        project: Callable[[Tensor], Tensor],
+    ):
```
Comment (Balandat) [inline botorch/acquisition/knowledge_gradient.py:456]: Implied, but our convention is to add hints throughout.

```suggestion
    ) -> None:
```

File: botorch/acquisition/knowledge_gradient.py [hunk 4]
```diff
@@ -428,16 +443,44 @@ def forward(self, X: Tensor) -> Tensor:
...
+        base_value_function = qSimpleRegret(
+            model=model, sampler=sampler, objective=objective
+        )
+    else:
+        base_value_function = PosteriorMean(model=model, objective=objective)
+    if project is None:
+        return base_value_function
     else:
-        return PosteriorMean(model=model, objective=objective)
+        return ProjectedValueFunction(base_value_function, project)
```
Comment (Balandat) [inline botorch/acquisition/knowledge_gradient.py:483]: Our convention is to call with explicit kwargs everywhere (where it makes sense...)

```suggestion
        return ProjectedValueFunction(
            base_value_function=base_value_function,
            project=project,
        )
```

File: botorch/acquisition/knowledge_gradient.py [hunk 5]
```diff
@@ -428,16 +443,44 @@ def forward(self, X: Tensor) -> Tensor:
         return values.mean(dim=0)
 
 
+class ProjectedValueFunction(AcquisitionFunction):
```
Comment (Balandat) [inline botorch/acquisition/knowledge_gradient.py:446]: Let's make sure names are as descriptive as possible. I added clarifying note in the docstring that this is typically used for value functions.

```suggestion
class ProjectedAcquisitionFunctionFunction(AcquisitionFunction):
```

File: test/acquisition/test_knowledge_gradient.py [hunk 6]
```diff
@@ -399,6 +400,62 @@ def test_evaluate_q_multi_fidelity_knowledge_gradient(self):
...
+                                torch.ones(1, 1, 1, device=self.device, dtype=dtype),
+                                torch.ones(1, device=self.device, dtype=dtype),
+                            ),
+                        ):
+                            val = qMFKG.evaluate(
+                                X=torch.zeros(1, 1, 1, device=self.device, dtype=dtype),
+                                bounds=torch.tensor([[0.0], [1.0]]),
+                                num_restarts=1,
+                                raw_samples=1,
+                            )
```
Comment (Balandat) [inline test/acquisition/test_knowledge_gradient.py:446]: Can use `ExitStack` to get rid of the nested contexts

```suggestion
            with ExitStack() as es:
                patch_f = es.enter_context(
                    mock.patch.object(MockModel, "fantasize", return_value=mfm),
                )
                mock_num_outputs = es.enter_context(
                    mock.patch(NO, new_callable=mock.PropertyMock),
                )
                es.enter_context(
                    mock.patch(
                       "botorch.optim.optimize.optimize_acqf",
                       return_value=(
                           torch.ones(1, 1, 1, device=self.device, dtype=dtype),
                           torch.ones(1, device=self.device, dtype=dtype),
                       ),
                   ),
                )
                es.enter_context(
                    mock.patch(
                        "botorch.generation.gen.gen_candidates_scipy",
                        return_value=(
                            torch.ones(1, 1, 1, device=self.device, dtype=dtype),
                            torch.ones(1, device=self.device, dtype=dtype),
                        ),
                    ),
                )

                mock_num_outputs.return_value = 1
                qMFKG = qMultiFidelityKnowledgeGradient(
                    model=mm,
                    num_fantasies=n_f,
                    X_pending=torch.rand(1, 1, 1, device=self.device, dtype=dtype),
                    current_value=torch.zeros(1, device=self.device, dtype=dtype),
                    cost_aware_utility=cau,
                    project=lambda X: torch.zeros_like(X),
                    expand=lambda X: torch.ones_like(X),
                )
                val = qMFKG.evaluate(
                     X=torch.zeros(1, 1, 1, device=self.device, dtype=dtype),
                     bounds=torch.tensor([[0.0], [1.0]]),
                     num_restarts=1,
                     raw_samples=1,
                 )
```

File: test/acquisition/test_knowledge_gradient.py [hunk 7]
```diff
@@ -399,6 +400,62 @@ def test_evaluate_q_multi_fidelity_knowledge_gradient(self):
...
+                    patch_f.asset_called_once()
+                    cargs, ckwargs = patch_f.call_args
+                    self.assertTrue(
+                        (
+                            ckwargs["X"]
+                            == torch.ones(2, 1, 1, device=self.device, dtype=dtype)
```
Comment (Balandat) [inline test/acquisition/test_knowledge_gradient.py:452]: use `torch.equal` here

**Review by Balandat — COMMENTED** [review]
Sorry, just got to this. Thanks for putting this up. A few nits mostly, o/w this looks good. 

@danielrjiang, @RaulAstudillo06 any comments / thoughts?

File: test/acquisition/test_knowledge_gradient.py (hunk 6 above)
Comment (saitcakmak) [reply test/acquisition/test_knowledge_gradient.py:446]: Cool! I've never used `ExitStack` before.

**Comment by saitcakmak** [comment]
Thanks for the review! I implemented the changes. Using `torch.equal` for the test revealed a small bug (concatenating pending points twice), which got fixed in the process. Turns out `torch_ones(2, 1, 1) == torch.ones(1, 3, 1)` returns a `2 x 3 x 1` tensor of `True` instead of raising an error.

File: test/acquisition/test_knowledge_gradient.py [hunk 8]
```diff
@@ -3,7 +3,7 @@
 #
 # This source code is licensed under the MIT license found in the
 # LICENSE file in the root directory of this source tree.
-
+from contextlib import ExitStack
```
Comment (Balandat) [inline test/acquisition/test_knowledge_gradient.py:6]: Mind adding an empty line here? Our linter is complaining about this. 

```suggestion

from contextlib import ExitStack
```

**Review by Balandat — CHANGES_REQUESTED** [review]

**Comment by Balandat** [comment]
Thanks for contributing and sorry about the delay merging this in.

**Comment by saitcakmak** [comment]
Thanks for the reviews. Happy to be contribute :)

**Pull request closed without merging**